proto/cosmos/bank/v1beta1/query.proto

```diff
@@ -7,18 +7,21 @@ import "google/api/annotations.proto";
 import "cosmos/base/v1beta1/coin.proto";
 import "cosmos/bank/v1beta1/bank.proto";
 import "cosmos_proto/cosmos.proto";
+import "cosmos/query/v1/query.proto";
 
 option go_package = "github.com/cosmos/cosmos-sdk/x/bank/types";
 
 // Query defines the gRPC querier service.
 service Query {
   // Balance queries the balance of a single coin for a single account.
   rpc Balance(QueryBalanceRequest) returns (QueryBalanceResponse) {
+    option (cosmos.query.v1.deterministic) = true;
     option (google.api.http).get = "/cosmos/bank/v1beta1/balances/{address}/by_denom";
   }
 
   // AllBalances queries the balance of all coins for a single account.
   rpc AllBalances(QueryAllBalancesRequest) returns (QueryAllBalancesResponse) {
+    option (cosmos.query.v1.deterministic) = true;
```

Comment:
This is true. But high gas costs, which smart contract devs should be aware of, is technically orthogonal to determinism. How do we express this to developers?

Reply:
Not sure.. the main point to communicate with the annotation should be "appropriate for queries from modules" which is both determinism + manageable gas.

I'm assuming this particular query does have proper gas tracking via the store so it would likely just run out of gas. But I do also question the usefulness of such an expensive query. I guess there is a scenario where a module might want to airdrop to all holders of a coin so that is a valid use case.

Reply:
@amaurym any thoughts?

Reply:
Fine to rename in this case. `internal` was proposed before, but i'm not a big fan as for client proto consumers that can be misleading. Maybe `inter_module`, `cross_module` or even `safe_for_state_machine`?

Reply:
`query_safe`?

Reply:
Note, re gas consumption, it's not a once size fits all. There might be chains or apps that are OK with using queries that consume a large amount of gas. What if we had two independent annotations?

Reply:
`module_query_safe`?

Reply:
As long as gas consumption is tracked and they're deterministic, expensive queries should be safe to use by modules. Not sure we need to annotate separately.

Reply:
Then I vote to keep it as `deterministic`

Reply:
There are 2 questions here:
1. Should we keep these high-gas queries in the state machine?
2. Should we rename the proto annotation?

IMO we can keep 1-yes 2-no, given Bez's arguments. To express the risk of high gas consumption, maybe we can simply add a proto comment on the gRPC query?

Reply:
There are 2 questions here:
1. Should we keep these high-gas queries in the state machine?
2. Should we rename the proto annotation?

IMO we can keep 1-yes 2-no, given Bez's arguments. To express the risk of high gas consumption, maybe we can simply add a proto comment on the gRPC query?

Reply:
ACK. Just leave it as-is and we can add warning comments saying gas _could_ be high.

Reply:
Thinking about this a bit more, I actually think we should rename. The three of us have the understanding regarding gas consumption, but we can't assume that about future users who might not read the docstring carefully. I think it is safer to use something like `module_query_safe` which prompts users to think more carefully about what "safe" means and if they don't know they're likely to check the doc string. Most people think they know what determinism means and then mark a query as such without thinking about other potential issues. I'm actually thinking of some concrete examples right now where I want to add queries around protobuf file descriptors which are probably deterministic but wouldn't track gas and I would want to mark them explicitly as `module_query_safe = false`.

Reply:
I just want to note again, that "high gas consumption" is relative to the chain. What one chain might consider high, another chain might deem as normal.

Reply:
Makes sense. Clearly the bigger issue is computation without gas consumption